File: concurrency/src/main/java/org/sqlunet/concurrency/ObservedDelegatingTask.java
package org.sqlunet.concurrency;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class ObservedDelegatingTask<Params, Progress extends Number, Result> extends Task<Params, Progress, Result>
{
	private final Task<Params, Progress, Result> delegate;

	private final TaskObserver.Observer<Progress> observer;

	public ObservedDelegatingTask(final Task<Params, Progress, Result> delegate, final TaskObserver.Observer<Progress> observer)
	{
		this.delegate = delegate;
		this.delegate.setForward(this);
		this.observer = observer;
	}

	@NonNull
	protected Task<Params, Progress, Result> getTask() { return this; }

	@Nullable
	@Override
	protected Result doInBackground(final Params[] params)
	{
		return this.delegate.doInBackground(params);
	}

	@Override
	protected void onPreExecute()
	{
		this.delegate.onPreExecute();
		this.observer.taskStart(this);
	}

	@Override
	protected void onPostExecute(final Result result)
	{
		this.delegate.onPostExecute(result);
		this.observer.taskFinish(true);
	}

	@Override
	protected void onCancelled(final Result result)
	{
		this.delegate.onCancelled(result);
		this.observer.taskFinish(false);
	}

	@Override
	protected void onCancelled()
	{
		this.delegate.onCancelled();
		this.observer.taskFinish(false);
	}

	@Override
	protected void onProgressUpdate(final Progress[] values)
	{
		this.delegate.onProgressUpdate(values);
		this.observer.taskUpdate(values[0], values[1]); // current, total
	}
}

File: concurrency/src/main/java/com/bbou/concurrency/ObservedDelegatingTask.kt
/*
 * Copyright (c) 2023. Bernard Bou <1313ou@gmail.com>.
 */
package com.bbou.concurrency

import com.bbou.concurrency.observe.TaskObserver

/**
 * Task that
 * -wraps a task and drives it, delegates work and call backs to that delegate
 * -hosts an observer fires events to its observer
 * The delegate task is set to forward posted (progress and result) messages to this wrapper.
 */
class ObservedDelegatingTask<Params, Progress : Pair<Number, Number>, Result>(private val delegateTask: Task<Params, Progress, Result>, observer: TaskObserver<Progress>) : Task<Params, Progress, Result>() {

    private val observer: TaskObserver<Progress>

    init {
        // termination and progress signals are forwarded to this delegating task
        delegateTask.setForward(this)
        this.observer = observer
    }

    private val task: Task<Params, Progress, Result>
        get() = this

    override fun doJob(params: Params?): Result {
        return delegateTask.doJob(params)
    }

    override fun onPreExecute() {
        delegateTask.onPreExecute()
        observer.taskStart(this)
    }

    override fun onDone(result: Result?) {
        delegateTask.onDone(result)
        observer.taskFinish(true)
    }

    override fun onCancelled() {
        delegateTask.onCancelled()
        observer.taskFinish(false)
    }

    override fun onProgress(progress: Progress) {
        delegateTask.onProgress(progress)
        observer.taskProgress(progress) // current, total
    }
}
